File: packages/ios/Sources/SeasonSprint/Networking/APIClient.swift
import Foundation
import ClerkKit

enum APIError: Error {
    case notAuthenticated
    case http(Int)
}

/// Thin REST client for the season-sprint server, authenticated with the Clerk
/// session JWT (the server accepts `Authorization: Bearer <jwt>` — see
/// packages/server/src/middleware/auth.ts).
enum APIClient {
    /// Current Clerk session JWT as an `Authorization` header value, or nil if signed out.
    @MainActor
    static func authorizationHeader() async -> String? {
        guard let session = Clerk.shared.session else { return nil }
        guard let jwt = try? await session.getToken() else { return nil }
        return "Bearer \(jwt)"
    }

    /// `GET /me/records?mode=...`
    @MainActor
    static func getRecords(mode: GameMode) async throws -> [APIRecord] {
        guard let auth = await authorizationHeader() else { throw APIError.notAuthenticated }
        var comps = URLComponents(url: Config.serverURL.appendingPathComponent("me/records"), resolvingAgainstBaseURL: false)!
        comps.queryItems = [URLQueryItem(name: "mode", value: mode.wireValue)]
        var req = URLRequest(url: comps.url!)
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: req)
        try check(response)
        return try JSONDecoder().decode([APIRecord].self, from: data)
    }

    /// `POST /me/records` — upsert a record by date; returns the saved record.
    @MainActor
    static func upsertRecord(date: String, winPoints: Int, mode: GameMode) async throws -> APIRecord {
        guard let auth = await authorizationHeader() else { throw APIError.notAuthenticated }
        var req = URLRequest(url: Config.serverURL.appendingPathComponent("me/records"))
        req.httpMethod = "POST"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(RecordWrite(date: date, winPoints: winPoints, mode: mode.wireValue))
        let (data, response) = try await URLSession.shared.data(for: req)
        try check(response)
        return try JSONDecoder().decode(UpsertResponse.self, from: data).record
    }

    /// `PUT /me/records/:id` — edit a record in place; returns the updated record.
    @MainActor
    static func updateRecord(id: String, date: String, winPoints: Int) async throws -> APIRecord {
        guard let auth = await authorizationHeader() else { throw APIError.notAuthenticated }
        var req = URLRequest(url: Config.serverURL.appendingPathComponent("me/records/\(id)"))
        req.httpMethod = "PUT"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(RecordWrite(date: date, winPoints: winPoints))
        let (data, response) = try await URLSession.shared.data(for: req)
        try check(response)
        return try JSONDecoder().decode(APIRecord.self, from: data)
    }

    /// `DELETE /me/records/:id`
    @MainActor
    static func deleteRecord(id: String) async throws {
        guard let auth = await authorizationHeader() else { throw APIError.notAuthenticated }
        var req = URLRequest(url: Config.serverURL.appendingPathComponent("me/records/\(id)"))
        req.httpMethod = "DELETE"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        let (_, response) = try await URLSession.shared.data(for: req)
        try check(response)
    }

    /// `GET /seasons` — public (no auth); returns the current season window, or nil.
    static func getSeasons() async throws -> Season? {
        try await getAllSeasons().currentSeason
    }

    /// `GET /seasons` — public (no auth); the full payload (every season + the current one).
    static func getAllSeasons() async throws -> SeasonsResponse {
        let req = URLRequest(url: Config.serverURL.appendingPathComponent("seasons"))
        let (data, response) = try await URLSession.shared.data(for: req)
        try check(response)
        return try seasonsDecoder.decode(SeasonsResponse.self, from: data)
    }

    /// Decoder that parses ISO-8601 timestamps with fractional seconds (e.g. `...000Z`).
    private static let seasonsDecoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { d in
            // Build formatters inside the closure: ISO8601DateFormatter isn't Sendable,
            // so it can't be captured by this @Sendable strategy closure.
            let withFractional = ISO8601DateFormatter()
            withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            let plain = ISO8601DateFormatter()
            plain.formatOptions = [.withInternetDateTime]

            let container = try d.singleValueContainer()
            let str = try container.decode(String.self)
            if let date = withFractional.date(from: str) ?? plain.date(from: str) {
                return date
            }
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Bad date: \(str)")
        }
        return decoder
    }()

    /// `POST /me/stream/token` — exchanges the Clerk auth for a short-lived SSE token.
    @MainActor
    static func getStreamToken() async throws -> String {
        guard let auth = await authorizationHeader() else { throw APIError.notAuthenticated }
        var req = URLRequest(url: Config.serverURL.appendingPathComponent("me/stream/token"))
        req.httpMethod = "POST"
        req.setValue(auth, forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: req)
        try check(response)
        return try JSONDecoder().decode(StreamTokenResponse.self, from: data).token
    }

    private static func check(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        guard (200..<300).contains(http.statusCode) else { throw APIError.http(http.statusCode) }
    }
}

File: packages/android/app/src/main/java/com/lcarthur/seasonsprint/data/ApiClient.kt
package com.lcarthur.seasonsprint.data

import com.lcarthur.seasonsprint.Config
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

/** Thrown when the server returns a non-2xx status. */
class ApiException(val code: Int) : Exception("HTTP $code")

/**
 * Thin REST client for the season-sprint server, mirroring iOS APIClient.swift.
 * The Clerk Bearer token is attached by [AuthInterceptor]; OkHttp is also reused for
 * the live WebSocket (Phase 7).
 */
object ApiClient {
    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMedia = "application/json".toMediaType()
    private val emptyBody = ByteArray(0).toRequestBody(null)

    val httpClient: OkHttpClient = OkHttpClient.Builder()
        .addInterceptor(AuthInterceptor())
        .build()

    private fun url(path: String) = "${Config.SERVER_URL}/$path"

    /** `GET /me/records` */
    suspend fun getRecords(): List<ApiRecord> = withContext(Dispatchers.IO) {
        val req = Request.Builder().url(url("me/records")).get().build()
        httpClient.newCall(req).execute().use { resp ->
            check(resp)
            json.decodeFromString<List<ApiRecord>>(resp.body.string())
        }
    }

    /** `POST /me/records` — upsert a record by date; returns the saved record. */
    suspend fun upsertRecord(date: String, winPoints: Int): ApiRecord = withContext(Dispatchers.IO) {
        val body = json.encodeToString(RecordWrite(date, winPoints)).toRequestBody(jsonMedia)
        val req = Request.Builder().url(url("me/records")).post(body).build()
        httpClient.newCall(req).execute().use { resp ->
            check(resp)
            json.decodeFromString<UpsertResponse>(resp.body.string()).record
        }
    }

    /** `PUT /me/records/:id` — edit a record in place; returns the updated record. */
    suspend fun updateRecord(id: String, date: String, winPoints: Int): ApiRecord = withContext(Dispatchers.IO) {
        val body = json.encodeToString(RecordWrite(date, winPoints)).toRequestBody(jsonMedia)
        val req = Request.Builder().url(url("me/records/$id")).put(body).build()
        httpClient.newCall(req).execute().use { resp ->
            check(resp)
            json.decodeFromString<ApiRecord>(resp.body.string())
        }
    }

    /** `DELETE /me/records/:id` */
    suspend fun deleteRecord(id: String): Unit = withContext(Dispatchers.IO) {
        val req = Request.Builder().url(url("me/records/$id")).delete().build()
        httpClient.newCall(req).execute().use { resp -> check(resp) }
    }

    /** `GET /seasons` — public (no auth); returns the current season window, or null. */
    suspend fun getSeasons(): SeasonDto? = withContext(Dispatchers.IO) {
        val req = Request.Builder().url(url("seasons")).get().build()
        httpClient.newCall(req).execute().use { resp ->
            check(resp)
            json.decodeFromString<SeasonsResponse>(resp.body.string()).currentSeason
        }
    }

    /** `POST /me/stream/token` — exchanges the Clerk auth for a short-lived WebSocket token. */
    suspend fun getStreamToken(): String = withContext(Dispatchers.IO) {
        val req = Request.Builder().url(url("me/stream/token")).post(emptyBody).build()
        httpClient.newCall(req).execute().use { resp ->
            check(resp)
            json.decodeFromString<StreamTokenResponse>(resp.body.string()).token
        }
    }

    private fun check(resp: Response) {
        if (!resp.isSuccessful) throw ApiException(resp.code)
    }
}
